webapp/src/ts/modules/tasks/tasks.component.ts

```diff
@@ -142,7 +155,29 @@ export class TasksComponent implements OnInit, OnDestroy {
       .then(taskDocs => {
         this.hasTasks = taskDocs.length > 0;
         this.loading = false;
-        this.tasksActions.setTasksList(this.hydrateEmissions(taskDocs));
+        this.hydrateEmissions(taskDocs)
+          .then((tasks) => {
+            // get lineages for all tasks
+            this.getLineagesFromTaskDocs(tasks)
+              .then((subjects) => {
+                const deepCopy = obj => JSON.parse(JSON.stringify(obj));
+                const lineagedTasks = deepCopy(tasks);
+                lineagedTasks?.forEach((task) => {
+                  // map tasks with lineages
+                  let lineage = _map(
```

Comment:
can you use native javascript functions for `map` and `find`?

Reply:
This can be a separate function, then be a step by step logic, that would make it easier to read. For example:
```
private getTaskLineage(subjects, task) {
  if (!subjects?.lenght || !this.currentLevel) { 
     return; // Exit early whenever possible!! Keeps code flat and understandable 
  }

 // Step by step logic to help readability & maintenance 
  const subject = subject.find( ...... find leanage .... );
  return lineage.map(...... extract what you need ....); 
}

private removeCurrentLineage(lineage) {
  if (!this.currentLevel || !lineage?.length) {
    return; // Exit early whenever possible!! Keeps code flat and understandable 
  }

  return lineage.filter(level => level && level !== this.currentLevel);
}
```

Reply:
Then in the `refreshTasks()` will look cleaner and flatter because we took out some code, we can restructure something like this, but test it!:

```
async refreshTasks() {
try {
   ...  Original telemetry code here ....

  const isEnabled = await this.rulesEngineService.isEnabled();
  this.tasksDisabled = !isEnabled;
  const taskDocs = isEnabled ? this.rulesEngineService.fetchTaskDocsForAllContacts() : [];

  this.hasTasks = taskDocs.length > 0;
  this.loading = false;

  const hydratedTasks = await this.hydrateEmissions(taskDocs) || [];
  const subjects = await this.getLineagesFromTaskDocs(hydratedTasks);

  hydratedTasks.forEach(task => {
    const lineage = getTaskLineage(task); // the new function
    task.lineage = removeCurrentLineage(lineage); // the other new function
  });

  this.tasksActions.setTasksList(hydratedTasks);

  if (!this.tasksLoaded) {
     this.tasksActions.setTasksLoaded(true);
  }

  ... telemetry code ...
  } catch (exception) {
 .... exception code....
 } 
}
```

Reply:
Yeah great suggestion,  thank you mvery much, I refactored the whole refreshTasks() function following your guidelines, you can take a look to the updated PR.